### ISSUE TITLE
Maximizing  a custom loss function

### ISSUE THREAD

**Issue by heaven00** [fchollet/keras#7249]
I want to maximize R^2 for my problem statement. I wrote a custom loss function for this based on https://en.wikipedia.org/wiki/Coefficient_of_determination
```
def nn_r2_score(y, prediction):
    total_error = tf.reduce_sum(tf.square(tf.subtract(y, tf.reduce_mean(y))))
    residual_error = tf.reduce_sum(tf.square(tf.subtract(y, prediction)))
    R_squared = tf.subtract(1.0, tf.div(total_error, residual_error))
    return R_squared
```

So far I haven't been able to find an option for this in the documentation or stackoverflow. Is there a workaround for it ?

**Comment by td2014** [comment]
@heaven00 .  When you say you are looking for a "workaround," what is the issue?  If you want to minimize the loss, you can always change the sign like the case of the cosine_proximity loss:
https://github.com/fchollet/keras/blob/master/keras/losses.py
Also, you should probably try to use the same pattern as in the source code, so that the shapes/dimensions are handled correctly.  I hope this helps.

**Comment by heaven00** [comment]
@td2014 thanks for the quick reply

```
def nn_r2_score(y, prediction):
    total_error = K.sum(K.square( y_true - K.mean(y_true) ) )
    residual_error = K.sum(K.square( y_true - y_pred ))
    R_squared = 1 - (residual_error / total_error)
    return -R_squared

```
did you meant something like the above ?

**Comment by td2014** [comment]
@heaven00 .  Yes.  You may need to experiment with the "axis=-1" (like in the code), to make sure you are computing the results over the dimensions you expected.  You can print the shape of the resulting tensors to check, but I think you have the right idea.  I hope it works out okay.

**Comment by heaven00** [comment]
Thanks, closing the issue :+1:
